File: Problem2.py
'''

Time Complexity : O(log (n))
Space Complexity : O(1)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No issues faced

Your code here along with comments explaining your approach

Since we are dealing with a BST, we need to look if both elements are on left/right. Or else, they are on either sides. So we are returning
the current root
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        cur = root
        while cur:
            if p.val > cur.val and q.val > cur.val:
                cur = cur.right
            elif p.val < cur.val and q.val < cur.val:
                cur = cur.left
            else:
                return cur

File: problem2.java
package s30.Trees-4;

public class problem2 {
    class Solution {
        //TC:- O(h)
        //TC:- O(h)
        public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
            if(root == null) return null;
            
            if(root.val > p.val && root.val > q.val){
                return lowestCommonAncestor(root.left,p,q);
            }else if(root.val < p.val && root.val < q.val){
                return lowestCommonAncestor(root.right,p,q);
            }else{
                return root;
            }
        }
    }
}
